File: src/potential/CentralPotential.cpp
#include "python.hpp"
#include "potential/CentralPotential.hpp"

using namespace espresso;
using namespace espresso::potential;

//////////////////////////////////////////////////
// REGISTRATION WITH PYTHON
//////////////////////////////////////////////////
real
CentralPotential::
computeEnergy(const Real3D dist) const {
  return this->computeEnergySqr(dist.sqr());
}

real
CentralPotential::
computeEnergy(const real dist) const {
  return this->computeEnergySqr(dist * dist);
}

namespace espresso {
  namespace potential {
    class _PythonCentralPotential
      : public python::wrapper< CentralPotential > {
    public:
      real _getCutoffSqr() const {
	return get_override("getCutoffSqr")();
      }
      
      real _computeEnergySqr(const real distSqr) const {
	return this->get_override("computeEnergySqr")(distSqr);
      }
      
      Real3D _computeForce(const Real3D dist) const {
	return get_override("computeForce")(dist);
      }
    };

    typedef CentralPotentialWrapper< _PythonCentralPotential > PythonCentralPotential;

  }
}

void
CentralPotential::registerPython() {
  using namespace espresso::python;

  real (CentralPotential::*computeEnergy1)(const Real3D) const =
    &CentralPotential::computeEnergy;
  real (CentralPotential::*computeEnergy2)(const real) const =
    &CentralPotential::computeEnergy;

  class_< espresso::potential::PythonCentralPotential, bases< Potential >, boost::noncopyable >
    ("potential_PythonCentralPotential")
    .def("computeEnergySqr", 
	 pure_virtual(&CentralPotential::computeEnergySqr))
    .def("computeEnergy", computeEnergy1)
    .def("computeEnergy", computeEnergy2)
    ;

}

File: src/potential/CentralPotential.py
from espresso import pmi
from espresso.potential.Potential import *

from _espresso import potential_PythonCentralPotential

class PythonCentralPotentialLocal(PotentialLocal, potential_PythonCentralPotential):
    def getCutoffSqr(self):
        pass

    def computeForce(self, *args):
        """Override this method to compute the force for a given distance.
        
        It should at least be able to handle a Real3D distance input.
        """
        pass

    def computeEnergySqr(self, distSqr):
        """Override this method to compute the energy at a given distance.
        
        It should at least be able to handle a Real3D distance input.
        """
        pass

    
class CentralPotentialLocal(PotentialLocal):
    def computeEnergy(self, *args):
        if len(args) == 1:
            arg0 = args[0]
            if type(arg0) is float or type(arg0) is int:
                return self.cxxclass.computeEnergy(self, arg0)
        return PotentialLocal.computeEnergy(self, *args)

CentralPotentialLocal = PythonCentralPotentialLocal

# from _espresso import potential_PythonCentralPotential
# class PythonCentralPotentialLocal(CentralPotentialLocal, potential_PythonCentralPotential):
#     pass

if pmi.IS_CONTROLLER:
    class CentralPotential(Potential):
        pmiproxydefs = Potential.pmiproxydefs
